// @google-cloud/translate/build/src/index.js
/*!
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const arrify = require("arrify");
const common_1 = require("@google-cloud/common");
const promisify_1 = require("@google-cloud/promisify");
const extend = require("extend");
const is = require("is");
const isHtml = require('is-html');
const teeny_request_1 = require("teeny-request");
const PKG = require('../../package.json');
/**
 * With [Google Translate](https://cloud.google.com/translate), you can
 * dynamically translate text between thousands of language pairs.
 *
 * The Google Translate API lets websites and programs integrate with Google
 * Translate programmatically.
 *
 * @class
 *
 * @see [Getting Started]{@link https://cloud.google.com/translate/v2/getting_started}
 * @see [Identifying your application to Google]{@link https://cloud.google.com/translate/v2/using_rest#auth}
 *
 * @param {ClientConfig} [options] Configuration options.
 *
 * @example
 * //-
 * // <h3>Custom Translate API</h3>
 * //
 * // The environment variable, `GOOGLE_CLOUD_TRANSLATE_ENDPOINT`, is honored as
 * // a custom backend which our library will send requests to.
 * //-
 *
 * @example <caption>include:samples/quickstart.js</caption>
 * region_tag:translate_quickstart
 * Full quickstart example:
 */
class Translate extends common_1.Service {
    constructor(options) {
        let baseUrl = 'https://translation.googleapis.com/language/translate/v2';
        if (process.env.GOOGLE_CLOUD_TRANSLATE_ENDPOINT) {
            baseUrl = process.env.GOOGLE_CLOUD_TRANSLATE_ENDPOINT.replace(/\/+$/, '');
        }
        const config = {
            baseUrl,
            scopes: ['https://www.googleapis.com/auth/cloud-platform'],
            packageJson: require('../../package.json'),
            projectIdRequired: false,
            requestModule: teeny_request_1.teenyRequest,
        };
        super(config, options);
        this.options = options || {};
        this.options.request = config.requestModule;
        if (this.options.key) {
            this.key = this.options.key;
        }
    }
    detect(input, callback) {
        const inputIsArray = Array.isArray(input);
        input = arrify(input);
        this.request({
            method: 'POST',
            uri: '/detect',
            json: {
                q: input,
            },
        }, (err, resp) => {
            if (err) {
                callback(err, null, resp);
                return;
            }
            let results = resp.data.detections.map((detection, index) => {
                const result = extend({}, detection[0], {
                    input: input[index],
                });
                // Deprecated.
                // tslint:disable-next-line no-any
                delete result.isReliable;
                return result;
            });
            if (input.length === 1 && !inputIsArray) {
                results = results[0];
            }
            callback(null, results, resp);
        });
    }
    getLanguages(targetOrCallback, callback) {
        let target;
        if (is.fn(targetOrCallback)) {
            callback = targetOrCallback;
            target = 'en';
        }
        else {
            target = targetOrCallback;
        }
        const reqOpts = {
            uri: '/languages',
            useQuerystring: true,
            qs: {},
        };
        if (target && is.string(target)) {
            reqOpts.qs.target = target;
        }
        this.request(reqOpts, (err, resp) => {
            if (err) {
                callback(err, null, resp);
                return;
            }
            const languages = resp.data.languages.map((language) => {
                return {
                    code: language.language,
                    name: language.name,
                };
            });
            callback(null, languages, resp);
        });
    }
    translate(inputs, optionsOrTo, callback) {
        const inputIsArray = Array.isArray(inputs);
        const input = arrify(inputs);
        let options = {};
        if (typeof optionsOrTo === 'object') {
            options = optionsOrTo;
        }
        else if (typeof optionsOrTo === 'string') {
            options = { to: optionsOrTo };
        }
        // tslint:disable-next-line no-any
        const body = {
            q: input,
            format: options.format || (isHtml(input[0]) ? 'html' : 'text'),
        };
        if (is.string(options)) {
            body.target = options;
        }
        else {
            if (options.from) {
                body.source = options.from;
            }
            if (options.to) {
                body.target = options.to;
            }
            if (options.model) {
                body.model = options.model;
            }
        }
        if (!body.target) {
            throw new Error('A target language is required to perform a translation.');
        }
        this.request({
            method: 'POST',
            uri: '',
            json: body,
        }, (err, resp) => {
            if (err) {
                callback(err, null, resp);
                return;
            }
            let translations = resp.data.translations.map((x) => x.translatedText);
            if (body.q.length === 1 && !inputIsArray) {
                translations = translations[0];
            }
            callback(err, translations, resp);
        });
    }
    request(reqOpts, callback) {
        if (!this.key) {
            super.request(reqOpts, callback);
            return;
        }
        reqOpts.uri = this.baseUrl + reqOpts.uri;
        reqOpts = extend(true, {}, reqOpts, {
            qs: {
                key: this.key,
            },
            headers: {
                'User-Agent': common_1.util.getUserAgentFromPackageJson(PKG),
            },
        });
        common_1.util.makeRequest(reqOpts, this.options, callback);
    }
}
exports.Translate = Translate;
/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
promisify_1.promisifyAll(Translate, { exclude: ['request'] });
